ctapipe/image/ellipsoid.py

```diff
@@ -0,0 +1,407 @@
+    """
+    # For likelihood calculation we need the width of the
+    # pedestal distribution for each pixel
+    # currently this is not available from the calibration,
+    # so for now lets hard code it in a dict
+
+    pedestal = PED_TABLE[geom.name]
+    pdf = PDFType(pdf)
+    pdf_dict = {
+        PDFType.gaussian: SkewedGaussian,
+        PDFType.skewed: SkewedGaussian,
+    }
+
+    unit = geom.pix_x.unit
+    pix_x = geom.pix_x
+    pix_y = geom.pix_y
+    image = np.asanyarray(image, dtype=np.float64)
+
+    if np.sum(image) == 0.0:
+        raise ImageFitParameterizationError("size=0, cannot calculate HillasParameters")
+
+    if isinstance(image, np.ma.masked_array):
+        image = np.ma.filled(image, 0)
+
+    if not (pix_x.shape == pix_y.shape == image.shape == cleaned_mask.shape):
+        raise ValueError("Image length and number of pixels do not match")
+
+    cleaned_image = image.copy()
+    cleaned_image[~cleaned_mask] = 0.0
+    cleaned_image[cleaned_image < 0] = 0.0
+
+    dilated_mask = cleaned_mask.copy()
+    for row in range(n_row):
+        dilated_mask = dilate(geom, dilated_mask)
+
+    dilated_image = image.copy()
+    dilated_image[~dilated_mask] = 0.0
+    dilated_image[dilated_image < 0] = 0.0
+    size = np.sum(dilated_image)
+
+    hillas = hillas_parameters(geom, cleaned_image)
+
+    x0 = create_initial_guess(geom, hillas, size)  # seeds
+
+    if np.count_nonzero(image) <= len(x0):
+        raise ImageFitParameterizationError(
+            "The number of free parameters is higher than the number of pixels to fit, cannot perform fit"
+        )
+
+    def likelihood(cog_x, cog_y, psi, length, width, skewness, amplitude):
+        parameters = [
+            cog_x * unit,
+            cog_y * unit,
+            length * unit,
+            width * unit,
+            psi * u.rad,
+            skewness,
+            amplitude,
+        ]
+
+        prediction = pdf_dict[pdf](*parameters).pdf(geom.pix_x, geom.pix_y)
+
+        prediction[np.isnan(prediction)] = 1e9
+        like = neg_log_likelihood_approx(dilated_image, prediction, SPE_WIDTH, pedestal)
+        if np.isnan(like):
+            like = 1e9
+        return like
+
+    if pdf == PDFType.gaussian:
+        x0["skewness"] = 0
+
+    m = Minuit(
+        likelihood,
+        **x0,
+    )
+
+    if pdf == PDFType.gaussian:
+        m.fixed = np.zeros(len(x0.values()))
+        m.fixed[-2] = True
+
+    if bounds is None:
+        bounds = boundaries(geom, image, dilated_mask, hillas, pdf)
+        m.limits = bounds
+    else:
+        m.limits = bounds
+
+    m.errordef = 1  # neg log likelihood
+    m.simplex().migrad()
+    m.hesse()
+
+    likelihood = m.fval
+    pars = m.values
+    errors = m.errors
+
+    like_array = likelihood
+    like_array *= dilated_mask
+    goodness_of_fit = np.sum(
+        like_array[like_array > 0]
+        - mean_poisson_likelihood_gaussian(
+            pdf_dict[pdf](
+                pars[0] * unit,
```

Comment:
This won't work for different kinds of PDFs that have different numbers of parameters, does it?